### PR TITLE
IGNITE-12366: Cancel file transmission on a node-receiver

### PR DESCRIPTION
**Pull request by Mmuzaf**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/managers/communication/TransmissionCancelException.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+package org.apache.ignite.internal.managers.communication;
+
+/**
+ * Exception is used to cancel a file transmission operation on the receiver. This exception may be thrown
+ * at anytime during session handling by a {@link TransmissionHandler} to gracefully interrupt the transmission
+ * session on a node-sender.
+ */
+public class TransmissionCancelException extends RuntimeException {
```
Comment (xtern) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/communication/TransmissionCancelException.java]: I think that TransmissionCancelledException sounds better.

Comment (xtern) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/communication/TransmissionCancelException.java]: Also, In my opinion, it's better to inherit this exception from the IgniteException.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/communication/GridIoManager.java [hunk 2]
```diff
@@ -3289,8 +3289,15 @@ public void send(
...
 
-                throw new IgniteCheckedException("Exception while sending file [rmtId=" + rmtId +
-                    ", file=" + file.getName() + ", sesKey=" + sesKey + ", retries=" + retries + ']', e);
+                if (X.hasCause(e, TransmissionCancelException.class)) {
+                    throw new TransmissionCancelException("File transmission has been cancelled on the remote node " +
+                        "[rmtId=" + rmtId + ", file=" + file.getName() + ", sesKey=" + sesKey +
+                        ", retries=" + retries + ", cause='" + e.getCause().getMessage() + "']");
+                }
+                else {
+                    throw new IgniteCheckedException("Exception while sending file [rmtId=" + rmtId +
```
Comment (xtern) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/communication/GridIoManager.java]: Why not just re-throw `e` here?

File: modules/core/src/main/java/org/apache/ignite/internal/managers/communication/GridIoManager.java [hunk 3]
```diff
@@ -3289,8 +3289,15 @@ public void send(
             catch (IgniteCheckedException e) {
                 closeChannelQuiet();
 
-                throw new IgniteCheckedException("Exception while sending file [rmtId=" + rmtId +
-                    ", file=" + file.getName() + ", sesKey=" + sesKey + ", retries=" + retries + ']', e);
+                if (X.hasCause(e, TransmissionCancelException.class)) {
+                    throw new TransmissionCancelException("File transmission has been cancelled on the remote node " +
+                        "[rmtId=" + rmtId + ", file=" + file.getName() + ", sesKey=" + sesKey +
+                        ", retries=" + retries + ", cause='" + e.getCause().getMessage() + "']");
```
Comment (xtern) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/communication/GridIoManager.java]: It seems that `e.getCause()`may not always contain expected exception (TransmissionCancelException).

File: modules/core/src/main/java/org/apache/ignite/internal/managers/communication/TransmissionCancelException.java (hunk 1 above)
Comment (Mmuzaf) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/communication/TransmissionCancelException.java]: Fixed.

Comment (Mmuzaf) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/communication/TransmissionCancelException.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/communication/GridIoManager.java (hunk 2 above)
Comment (Mmuzaf) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/communication/GridIoManager.java]: Fixed.

**Pull request merged**